### PR TITLE
Replaced deprecated Gtk::Alignment class using Widget margins

### PR DESCRIPTION
**Pull request by DhairyaBahl**
@rodolforg @Keyikedalube

After reading about alignment.h , I think that these are the most suitable changes ! If these are correct, then allow me to make these changes wherever possible ! 

Thanks [^_^]

### CODE REVIEW

**Comment by Keyikedalube** [comment]
Yeah, `set_halign()` and `set_valign()` are correct :)

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (rodolforg) [inline synfig-studio/src/gui/widgets/widget_defaults.cpp:199]: For consistency, please rename it without the prepended `_`. :)
And for better reading, without guessing, `outline` and not `otln` :P

**Review by rodolforg — CHANGES_REQUESTED** [review]

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (rodolforg) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp:199]: And please rephrase your commit ;)

Comment (DhairyaBahl) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp]: Renamed the widget_otln to outline ! And also rephrased the latest commit and headline of the issue ! Now resolving other widgets as well ! [^_^]

Comment (rodolforg) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp]: I meant  "removing alignment test" ;) -> "remove deprecated Gtk::Alignment" would be better

**Comment by DhairyaBahl** [comment]
Kindly tell me If I have to make any changes in this ! I think I successfully removed alignment.h from defaults.h file. If everything is alright till now, I can move forward after removing the alignment.h from defaults.cpp as well !

**Comment by rodolforg** [comment]
That's the way it is :)
But look your last commit message: it says it would remove the gtkmm/alignment.h , but it still didn't ;)

**Comment by DhairyaBahl** [comment]
> That's the way it is :)
> But look your last commit message: it says it would remove the gtkmm/alignment.h , but it still didn't ;)

I meant that defaults.h doesn't need alignment.h but still there are few buttons in defaults.cpp file that are still using alignment.h ! After removing that I will be done with alignment.h ! Then we can move ahead ! :)

**Comment by rodolforg** [comment]
OK, keep removing all explicit Gtk::Alignment usages in this PR :)
We'll squash them all into a single commit before merging this PR ;)

**Comment by DhairyaBahl** [comment]
@rodolforg @Keyikedalube I replaced alignment.h with fixed.h in 2 files ! Is that okay ? or should I use some better alternative for the same ? I guess this should work perfectly ! Kindly confirm it once ! So, that I can start replacing it further even outside the widgets i.e. in states, docks and even in the gui folder !

Thanks !

File: synfig-studio/src/gui/canvasview.cpp
Comment (rodolforg) [inline synfig-studio/src/gui/canvasview.cpp:1010]: Wrong indentation :/

File: synfig-studio/src/gui/canvasview.cpp
Comment (rodolforg) [inline synfig-studio/src/gui/canvasview.cpp:580]: If a widget is really necessary, what about just a plain Widget?

**Review by rodolforg — CHANGES_REQUESTED** [review]

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (rodolforg) [inline synfig-studio/src/gui/widgets/widget_defaults.cpp:249]: Is it really necessary? Why don't set valign/halign of _button_reset like you did in previous commit?

**Review by rodolforg — CHANGES_REQUESTED** [review]

File: synfig-studio/src/gui/canvasview.cpp
Comment (Keyikedalube) [reply synfig-studio/src/gui/canvasview.cpp:580]: We have `set_margin_start()` and `set_margin_end()` to achieve the same behavior. So instead of placing that space widget, find out where it is placed and determine if our next widget is proceeding or succeeding from space widget, then use the appropriate margin property.

File: synfig-studio/src/gui/canvasview.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/canvasview.cpp]: Fixed Indentation :)

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp]: I made the requested changes now ! kindly look into it

File: synfig-studio/src/gui/canvasview.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/canvasview.cpp:580]: I used Gtk::Fixed because I was thinking that it is the purpose of Fixed is to showcase that something related to arrangement is going on in the code which contains this class! If its not good, then I will change that to whatever needed ! No issues :)

Comment (rodolforg) [reply synfig-studio/src/gui/canvasview.cpp:580]: `Fixed` is a container that places children in fixed coordinates (they don't change if you expand/shrink the container dimensions).

**Comment by DhairyaBahl** [comment]
@rodolforg @Keyikedalube If changes till now are good to go, then We can work on new class or even on the same class. This is because it is getting quite messy to manage all the commits.

Thanks :)

File: synfig-studio/src/gui/canvasview.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/canvasview.cpp:580]: @rodolforg I am unable to find where canvaspreview is ? I want to try gtkmm container widget classes on it ! and it can't be done without trial and error (atleast for me 🙂). It would be great if you guys can tell me that how can I find some piece of code in synfig studio. As it will also help in the later stages as well !

Thanks and sorry for inconvenience. I will now keep on replacing alignment.h wherever there is no chance of conflicts for the time being.

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (Keyikedalube) [inline synfig-studio/src/gui/widgets/widget_defaults.cpp:241]: `Gtk::Misc` is deprecated plus I don't see any changes with the code. This class is properly styled with Synfig CSS.

**Review by Keyikedalube — CHANGES_REQUESTED** [review]

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (Keyikedalube) [inline synfig-studio/src/gui/widgets/widget_defaults.cpp:224]: This too... You can safely delete those lines and commit :)

**Review by Keyikedalube — CHANGES_REQUESTED** [review]

File: synfig-studio/src/gui/canvasview.cpp
Comment (Keyikedalube) [reply synfig-studio/src/gui/canvasview.cpp:580]: @DhairyaBahl `widget_space` is added here
https://github.com/synfig/synfig/blob/9b6c3a2a62796647f0937c6ad8354f1411c8776c/synfig-studio/src/gui/canvasview.cpp#L594-L596
You can notice it's trying to add 4 spaces between the canvas workspace and top/right toolbars. Delete the `widget_space` along with the `Gtk::Alignment` code and replace them with `widget_right_bar->set_margin_start(4)` (Check if this is really needed) and `widget_top_bar->set_margin_bottom(4)`. The ruler changes a bit if you try to use `widget_work_area->set_margin_top(4)` so I used top bar widget there.

You can use pen and paper to visually draw the layout determining how widgets are packed and finally decide which margin property to use `start/end/top/bottom`. Hope you got enough idea.

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp]: I have removed these in my latest commit :)

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp]: I have removed these in my latest commit :)

File: synfig-studio/src/gui/canvasview.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/canvasview.cpp]: I pushed my latest changes after making the required changes ! Thanks for helping with the grid ! after drawing it in my notebook, I was able to see a clear picture of the grid. 

I also wanted to ask how can I find particular code in the synfig-studio after building so that I can make changes and build code again and again and experiment with the code as much as possible. Kindly tell me how to do so. Sometimes I am able to find a part of code in synfig-studio and sometimes I fail to do so. Kindly tell me the best approach to do so.

Thanks and Regards :)

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (Keyikedalube) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp]: Resolved :)

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (Keyikedalube) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp]: Resolved :)

File: synfig-studio/src/gui/canvasview.cpp
Comment (Keyikedalube) [inline synfig-studio/src/gui/canvasview.cpp:587]: Figured this isn't really needed, just the work area widget will do because `widget_space` was added horizontally whereas the right toolbar is a vertical widget and there's no ruler to the right. So no extra spaces needed at all...

Comment (DhairyaBahl) [reply synfig-studio/src/gui/canvasview.cpp]: Got it ! made the required changes in the latest commit :)

File: synfig-studio/src/gui/canvasview.cpp
Comment (rodolforg) [reply synfig-studio/src/gui/canvasview.cpp]: No, it isn't ;) We use `tab` for source code identantion ;)

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (rodolforg) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp]: Please remove all these `Gtk::Fixed`. They are not meant for creating a 'fixed space'. If you want to force it with an empty widget, just use a `Gtk::Widget` ;)

Comment (rodolforg) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp]: If they contain a widget though, just apply the alignment info like you did for the fill-color widget.

File: synfig-studio/src/gui/canvasview.cpp
Comment (rodolforg) [reply synfig-studio/src/gui/canvasview.cpp]: > I also wanted to ask how can I find particular code in the synfig-studio after building so that I can make changes and build code again and again and experiment with the code as much as possible.

That depends on what you are talking about ;)
I listed some info about synfig code folder structure here:
https://forums.synfig.org/t/help-required-in-understanding-the-synfig-studio-code/10505/4

You can use `git grep` for searching some words inside project source code.

But if you are 'attacking' a specific file for now, I can't understand your question.

File: synfig-studio/src/gui/canvasview.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/canvasview.cpp]: I've fixed the indentation now ! just with 1 ``` tab ``` ! Kindly look into it :)

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp]: Got it ! Understood ! I will start replacing Gtk : :  Fixed and Gtk : : Alignment , wherever possible !

File: synfig-studio/src/gui/canvasview.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/canvasview.cpp]: Okkayy ! I got your point ! thanks for help !  Nope I meant the same but I was unable to frame the question lol ! my bad ! anyways thanks again

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp]: I guess I can't change name of the previous commit that I've already pushed ! I've tried renaming them and pushing them again but github won't show any change ! So, I will keep that in mind from next time [^_^]

**Comment by DhairyaBahl** [comment]
image

This is rendesc.cpp in src/gui/ . I wanted to confirm something ! padedPanel over here is being declared locally and being used again. So, I think It would be okay to just replace it with Gtk : : widget class ! and add the alignments using valign and halign ! If yes then I am on it , otherwise I'll have to look for another way.

image

Secondly, talking about all the deprecated set_alignment functions that are being used over here ! According to docs, set_alignment function is used on those widgets which are having Gtk : : Alignment as the child but I can't see any child of that type being added in these ! Kindly confirm this also ! if this is the case then we can remove these right away. 

Thanks :)

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (Keyikedalube) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp]: Instead of using `Gtk::Widget` as a replacement for `Gtk::Fixed` there's `set_margin_start/end/top/bottom()` to achieve this.

I'll requote my previous comment again here:
> We have set_margin_start() and set_margin_end() to achieve the same behavior. So instead of placing that space widget, find out where it is placed and determine if our next widget is preceding or succeeding from space widget, then use the appropriate margin property.

**Comment by Keyikedalube** [comment]
@DhairyaBahl 

> I wanted to confirm something ! padedPanel over here is being declared locally and being used again and again . So, I think It would be okay to just replace it with Gtk : : widget class

You can remove `Gtk::Alignment` code and return the second top container `Gtk::Box` instead.

> Secondly, talking about all the deprecated set_alignment functions that are being used over here ! According to docs, set_alignment function is used on those widgets which are having Gtk : : Alignment as the child but I can't see any child of that type being added in these ! Kindly confirm this also ! if this is the case then we can remove these right away.

That means [`set_alignment()`](https://developer.gnome.org/gtkmm/3.24/classGtk_1_1Entry.html#a0bf8a000d06a5754d734ea2b0f1a424e) is a member function of `Gtk::Entry` not `Gtk::Alignment`.
And no, there's no reason to remove them. They're needed for Right-to-Left text positioning.

File: synfig-studio/src/gui/widgets/widget_defaults.cpp
Comment (Keyikedalube) [reply synfig-studio/src/gui/widgets/widget_defaults.cpp]: BTW, there's nothing wrong with using `Gtk::Widget` as a replacement for `Gtk::Fixed`. It's still a solution though.

**Comment by DhairyaBahl** [comment]
@rodolforg @Keyikedalube Kindly tell me how to add xscale and yscale properties to the Gtk::box as I am unable to find how to do so ! except from that , I think that I've done everything correctly ! Kindly confirm the changes in rendesc.cpp so that then I can move forward !

**Comment by rodolforg** [comment]
It's written in the docs: https://developer.gnome.org/gtk3/stable/GtkAlignment.html#GtkAlignment--xscale

Use hexpand / vexpand on child widget

**Comment by DhairyaBahl** [comment]
@rodolforg @Keyikedalube 

Kindly check changes I've done till now ! Coz now I am only left with states ! 

image

and kindly tell me how exactly do I have to replace gtk : : alignment over here ! I can only think of one way , i.e. to replace the same with some widget ! coz some sort of widget is required over here for SPACING to work ! If there is a better way than this one, than kindly tell me about that and I will do that accordingly !

Also tell me in which folders do I have to look in after successful replacement of alignment.h from the synfig's code ? 

Thanks and Regards :)

File: synfig-studio/src/gui/renddesc.cpp
Comment (Keyikedalube) [inline synfig-studio/src/gui/renddesc.cpp:827]: If set margin value is 0 no need to add them. It's redundant. By default, it was already 0.

**Comment by Keyikedalube** [comment]
@DhairyaBahl You can implement rodolforg's `Gtk::Widget` suggestion for empty spaces.

**Comment by ice0** [comment]
> @DhairyaBahl You can implement rodolforg's `Gtk::Widget` suggestion for empty spaces.

I think better to use margins to set spacings as @Keyikedalube suggested already.

@DhairyaBahl Are you using GTK Inspector in your process? It helps a lot.

File: synfig-studio/src/gui/preview.cpp
Comment (ice0) [inline synfig-studio/src/gui/preview.cpp:459]: No need to use `Fixed` here. You can just remove it.

This is `Preview` window toolbar, you can open this window using `File->Preview-> (click Preview button)`

File: synfig-studio/src/gui/renddesc.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/renddesc.cpp:827]: I will fix this asap, wherever I have done this ! :)

File: synfig-studio/src/gui/preview.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/preview.cpp:459]: @ice0 Thank You ! but how will I come to know it myself that which file is meant for what ? i.e. how did you came to know that preview.cpp is for preview window toolbar in ``` file->preview->previewButton ``` ?  it would  help a lot if I can see these changes that I am making ! that's why I am having so much doubts.

and thank you for telling me about gtk inspector ! Its great ! Trying to get hang of that tool :)

Thank You So Much  [^_^]

Comment (ice0) [reply synfig-studio/src/gui/preview.cpp:459]: I searched for usages of the `Preview` class and found that it is being used from `Dialog_Preview::set_preview(etl::handle<Preview> prev)`. After that you need to search for usages of `Dialog_Preview` etc. But at that point I did not do any further search because I already knew where the preview dialog was (this is just an experience).

Comment (DhairyaBahl) [reply synfig-studio/src/gui/preview.cpp:459]: Understood ! Thank You so much for help ^_^

File: synfig-studio/src/gui/renddesc.cpp
Comment (DhairyaBahl) [reply synfig-studio/src/gui/renddesc.cpp]: I've made the required changes :)

**Comment by DhairyaBahl** [comment]
@Keyikedalube @rodolforg @ice0 Kindly tell me if I should start working on spacing right away or I should start a new PR for the same ? Kindly tell me so that I can start working on the same right away :)

**Comment by Keyikedalube** [comment]
> P.S. I propose to make another PR for `SPACING` stuff.

@DhairyaBahl better stick to ice0's plan

**Comment by DhairyaBahl** [comment]
@Keyikedalube @rodolforg I think this is it for the PR ! You can check this out and I can make changes if required ! Otherwise kindly merge this pr !

Secondly, I m sorry for being inactive from past two days coz I was having fever due to covid vaccination ! I will be regular from now on :) 

Thanks and regards :)

**Review by ice0 — APPROVED** [review]
Looks good to me. Looking forward to @rodolforg approval.

**Comment by ice0** [comment]
> Secondly, I m sorry for being inactive from past two days coz I was having fever due to covid vaccination

There is no problem here. Don't forget to take a break if you need them.

**Comment by ice0** [comment]
ping @rodolforg

**Comment by DhairyaBahl** [comment]
@ice0 Kindly merge this PR ! Then I will be able to check if alignment.h is left or not ! Actually I got confused between multiple git branches. So kindly merge this, if this PR is good to go.

Thanks :)

PS : Should I start the migration in app.cpp ? if not then kindly tell me on which section I now have to work ? Thanks again :)

**Comment by ice0** [comment]
Merged. Thank you!

**Comment by ice0** [comment]
> PS : Should I start the migration in app.cpp ? if not then kindly tell me on which section I now have to work ? Thanks again :)

Just trying to compile Synfig with deprecated classes disabled. These are top errors: 
```
Gtk::IconFactory
Gtk::ActionGroup
Gtk::IconFactory
Gtk::StockID
Gtk::ImageMenuItem
Gtk::ToggleAction
```
etc. I think better to start with one of them. In this case, you can easily check the result of your work.

**Comment by rodolforg** [comment]
@ice0 @DhairyaBahl First, sorry my absence: real life work caught me hard.

Replacing `Gtk::ImageMenuItem` will affect a litte the GUI: Gtk4 doesn't natively uses icons in menu anymore, as Gnome UI guidelines think they're a bad habit nowadays. I particularly don't know if I agree as they help user to remember/associate button icons to menu actions . However, button icons should be clear by themselves anyway :P

```
Gtk::ActionGroup
Gtk::StockID
Gtk::ToggleAction
Gtk::IconFactory
```

I think they will depend on Gtk::Application class… A long time ago, I started to try to remove StockID and replace IconFactory with IconTheme (https://github.com/synfig/synfig/pull/1508), but I stuck due to the extensive usage of now deprecated Gtk::Action, that is not supported by Gtk::Application.
The action concept changed a bit to allow, for example, to be called via DBus, and all GUI info (icon, tooltip, ...) for action is removed by GTK/GDK and, if needed, should be implemented by application.

**Comment by DhairyaBahl** [comment]
@rodolforg So, as I can observe, almost every class is now dependent on Gtk::Action ! So, should I start with replacement of Gtk::Action to Gio::Action first ? and then only we will be able to port every class to latest one.

**Comment by rodolforg** [comment]
No. Gio::Action 'only' works for Gtk::Application and Gtk::ApplicationWindow AFAIK

**Comment by DhairyaBahl** [comment]
@rodolforg I was talking about this. i.e. gtkmm ToggleAction. and secondly, you told me that replacement of stockID was stopped earlier because of Gtk::Action ! So, should I start replacing Gtk::Action first ?? 

image

Same description for Gtk : : ActionGroup as well ``` to use gio::action ``` ! It would be great if you can tell me which class should I pick first. 

Thank You :)

**Comment by rodolforg** [comment]
AFAIK we can't use Gio::Action feature without Gtk::Application (or Glib::Application).

**Comment by ice0** [comment]
@rodolforg Nice to see you again and hope that you are all right.

> Replacing `Gtk::ImageMenuItem` will affect a litte the GUI: Gtk4 doesn't natively uses icons in menu anymore, as Gnome UI guidelines think they're a bad habit nowadays. I particularly don't know if I agree as they help user to remember/associate button icons to menu actions . However, button icons should be clear by themselves anyway :P

I thought they deprecated this widget, just because you can do the same using GtkImage+GtkLabel combination.
> If you want to display an icon in a menu item, you should use GtkMenuItem and pack a GtkBox with a GtkImage and a GtkLabel instead. 

https://developer.gnome.org/gtk3/stable/GtkImageMenuItem.html


> AFAIK we can't use Gio::Action feature without Gtk::Application (or Glib::Application).

I didn't find a direct requirement for this, but the documentation says: 

> This interface provides a convenient way of associating widgets with actions on a GtkApplicationWindow or GtkApplication.

https://developer.gnome.org/gtk3/stable/GtkActionable.html

so I think you are right.

**Comment by rodolforg** [comment]
> I thought they deprecated this widget, just because you can do the same using GtkImage+GtkLabel combination.

Yes, we can do it this way, but we should use a placeholder for those menu items that don't have an icon to keep everything aligned.

Anyway, GTK is still somewhat aligned to Gnome Guidelines. And they don't like classic menus and Icons+Text combo (in buttons, menus, etc.) https://developer.gnome.org/hig/stable/

**Comment by DhairyaBahl** [comment]
@rodolforg @ice0 I am extremely sorry for being inactive from past 3 days ! This is because i was having college exams and along with that I was also not well ! I will now work on remaining classes. 

Sorry and Thanks :)

**Comment by rodolforg** [comment]
@DhairyaBahl any news?

**Comment by DhairyaBahl** [comment]
@rodolforg Sorry once again ! now I will be able to work full time again on synfig ! Kindly tell me that should I continue removing classes like I was doing before , if yes then it would be great if you can tell me which one should i pick first out of 

```
Gtk::ActionGroup
Gtk::StockID
Gtk::ToggleAction
Gtk::IconFactory
```

otherwise, I can start directly with app.cpp ! Kindly clear this small doubt and then I will be able to pace up myself ! 

Thanks and regards ! Sorry again :)

**Comment by ice0** [comment]
@DhairyaBahl 

As Rodolfo already said:

> AFAIK we can't use Gio::Action feature without Gtk::Application (or Glib::Application).

better to start from Gtk::Application (or Glib::Application)

**Comment by DhairyaBahl** [comment]
@ice0 Understood ! Starting working on it right away ! :) Thanks .

**Comment by DhairyaBahl** [comment]
@rodolforg @ice0 

Status Update : I found that there are few instances of ``` gtkmm/UImanager ``` and we can replace it with ``` gtkmm/builder ``` 

secondly, I was getting started with migration of window to application window but even a small change is breaking the app ! I tried the whole day but I couldn't find a starting point from where I can begin migration ! 

Directly changing the main.cpp will definitely break synfig ! So , top down approach is not an option ! so we have to first migrate the children of app.cpp to gtk applicationwindow but they require gtk application support first ! So I am also stuck there ! 

It would be great if you guys can help me finding a starting point from where I can start the migration ! 

Thanks a lot ! Till then I am replacing UImanager with builder ! Will discard the changes later if it is not required ! 

Thanks 🙂

**Comment by rodolforg** [comment]
> better to start from Gtk::Application (or Glib::Application)

The first thing to do is migrate `Gtk::Main` to `Gtk::Application`, not `Gtk::ApplicationWindow`.

Did you read gtkmm tutorial and some HowDoI? IIRC I listed some of them to you. Anyway, here they are again:
https://developer.gnome.org/GtkApplication/
https://developer.gnome.org/gtkmm-tutorial/3.18/sec-basics-simple-example.html
https://developer.gnome.org/gtkmm-tutorial/3.18/sec-helloworld.html
https://developer.gnome.org/gtkmm-tutorial/3.22/sec-buildapp-trivial-app.html
https://developer.gnome.org/gtkmm-tutorial/3.18/sec-basics-simple-example.html

**Comment by rodolforg** [comment]
> I found that there are few instances of `gtkmm/UImanager` and we can replace it with `gtkmm/builder`

Yes it should be done, but it's impossible now. Deprecated UIManager deals with deprecated Gtk::Action and Gtk::ActionGroup.
AFAIK Gtk::Builder doesn't handle any Action.

**Comment by rodolforg** [comment]
>  IIRC I listed some of them to you.

I did :)
https://github.com/synfig/synfig/issues/2123#issuecomment-854252608
https://github.com/synfig/synfig/issues/2123#issuecomment-855375078

**Comment by DhairyaBahl** [comment]
> > I found that there are few instances of `gtkmm/UImanager` and we can replace it with `gtkmm/builder`
> 
> Yes it should be done, but it's impossible now. Deprecated UIManager deals with deprecated Gtk::Action and Gtk::ActionGroup.
> AFAIK Gtk::Builder doesn't handle any Action.

Understood ! So, I will just focus on Gtk::Application right now !

**Comment by DhairyaBahl** [comment]
> Did you read gtkmm tutorial and some HowDoI? IIRC I listed some of them to you. Anyway, here they are again:

@rodolforg 

I remember you sent these to me 😃 ! I even read all of them ! Not only reading but also I made few apps this by following this tutorial ! 

I only got a bit confused on where exactly should I begin in huge codebase of synfig ! Anyhow , no issues ! I will give those another read and will also try again with synfig's codebase ! 

Thanks for resharing those links 🙂

**Comment by DhairyaBahl** [comment]
@rodolforg @ice0 

Status Update : I've gone through all the resources again that you reshared yesterday ! I think I missed [this](https://developer.gnome.org/gtkmm-tutorial/3.22/sec-buildapp-trivial-app.html) one when I was going through these for the first time ! This has proven to be the most informative one till now for me ! 

Now just starting the migration ! I got the concepts of gtk : :  application ! And getting started with them now ! :) 

Thanks :)

**Comment by DhairyaBahl** [comment]
@rodolforg @ice0 

Status Update : I am almost done with major changes ig ! Just debugging my code coz it is currently showing lot of errors ! 
I think its going to take few more hours to debug my mistakes and then I will run some tests on the code and then another session of debugging if needed, and then I will make a new PR for the same ! 

Thanks and regards 🙂

**Pull request merged**
